README.md

```diff
@@ -584,44 +598,44 @@ def build(env):
     # and use this information for a new template.
     env.template("module_header.hpp.in", substitutions={"headers": headers})
 
-    # You can add metadata to the build log which then made
-    # available in the post_build step. This is like a dictionary.
-    env.add_metadata("include_path", "generated_folder/")
+    # Add values to a collector, all these are type checked
+    env.collect("::include_path", "repo/must_be_valid_path/", "repo/folder2/")
```

Comment:
So this would append to the collector at `current_repo:current_module:...`. So you would almost always have some more in the lines of `:other_module:...`.

Something to keep in mind is that this adds dependencies for all modules which should use the collector. For the build modules this means almost every other module. This is fine, but means, that this module must not do any work by itself, because otherwise you have a module which you can never disable.

As the build collectors are a very general thing, maybe be should have lbuild built-in modules and promote it to something like that?

Reply:
> Something to keep in mind is that this adds dependencies for all modules which should use the collector.

Hm, interesting loophole: You can depend on `:module` that contains the `::include_path` collector and then it's guaranteed to exists.
Or you don't depend on `:module` and instead only add something to the collector **if** it exists:
```py
if env.has_collector("::include_path"):
    env.collect("::include_path", ...)
```

This would be a valid (and kinda elegant) way of dealing with this dependency problem. I would add another function `env.collect_silent(...)` that just ignores this call if the collector doesn't exist.

Reply:
> This is fine, but means, that this module must not do any work by itself, because otherwise you have a module which you can never disable.

Hm, if you want to let the user decide between two or more post-step modules, then none of the modules in your repo must depend on this, yes.
I think this is a generic issue tho, and the only way to deal with this is to have a module do nothing (or only the common operation) and then have the "choice" modules depend on this module.

I guess the question is: should `env.collect()` fail silently by default. I think that would be a terrible idea, because misspellings of the key then also fail silently. On the other hand there is no good way to check if the key is valid if the module hasn't registered them, so `env.collect_silent()` is always *really* silent.

Reply:
> I think this is a generic issue tho, and the only way to deal with this is to have a module do nothing (or only the common operation) and then have the "choice" modules depend on this module.

Exactly, only that the user has to implement it that way. This is not enforced in any way. Maybe we need something along the lines of a best practice section?

> should `env.collect()` fail silently by default

No. The `env.collect_silent()` would be more of a `collect_maybe()`. I think this is to error prone.

Reply:
> more of a `collect_maybe()`

I've added a function (I already implemented the behavior before). I briefly considered building a "shadow" tree of collectors, since all modules have their prepare method called. Then at least I could still check for the name and type, however, this will break as soon as two modules declare different collectors under the same name, but both returns False. I guess lbuild could still check the name for that aspect tho.

I mean, if no module can consume this data, then the type check is irrelevant, no?

> Maybe we need something along the lines of a best practice section?

We need actual documentation, as well as actually documenting the FacadeObjects with docstrings. I mean I started the README as a hack, because I wanted *something* to dump the docs in, but 1000+ lines is not useful to anyone. The issue is that I'm still researching and learning the Python best-practices, so I've put it off for now. I'm mostly conce

Reply:
> this will break as soon as two modules declare different collectors under the same name

I mean this wouldn't just be bad, it would be malicious practice, since changing the *type* of collector depending on repo options would result in some terrible semantics. But then again, we really need a best pratices section.

Reply:
Ok, we already have this shadow tree actually, every node has two flags: `available` and `selected`, which map to the return value of `prepare` and the result of the dependency resolution. 

By default, the name resolution only takes available *and* selected modules into consideration, but for `env.collect_maybe` the name resolution can just check for available modules, and then just proceed like normal, since the collector object is available and it can collect these values, they will just not be accessible in any post-build step (since their name resolution again only check for selected collectors).

Reply:
I would just make this the default behavior of `env.collect()`. If the module want to depend on the collector module, it can just depend on the collector module. There is no need to encode this twice. It's not the job of the module that adds this data to worry about whether it gets used or not. Just that it is exists and is of the right type.

Reply:
done.